### ISSUE TITLE
Crash: Segmentation Fault

### ISSUE THREAD

**Issue by igorescobar** [jiangwenyuan/nuster#85]
The issue reported here #84 keeps happening.

```
[ALERT] 146/005344 (1) : Current worker #1 (7) exited with code 139 (Segmentation fault)
2020-05-26T01:53:44.661+01:00
[ALERT] 146/005344 (1) : exit-on-failure: killing every processes with SIGTERM
2020-05-26T01:53:44.661+01:00
[WARNING] 146/005344 (1) : All workers exited. Exiting... (139)
2020-05-26T01:53:44.661+01:00
[NOTICE] 146/005241 (1) : New worker #1 (8) forked
2020-05-26T01:52:42.068+01:00
[NOTICE] 146/005241 (1) : [nuster][cache] on, dict_size=524288000, data_size=524288000
2020-05-26T01:52:42.067+01:00
<133>May 26 00:52:41 nuster[1]: Proxy fe started.
2020-05-26T01:52:42.067+01:00
<133>May 26 00:52:41 nuster[1]: Proxy be started.
2020-05-26T01:52:42.067+01:00
[ALERT] 146/005131 (1) : Current worker #1 (8) exited with code 139 (Segmentation fault)
2020-05-26T01:51:31.046+01:00
[ALERT] 146/005131 (1) : exit-on-failure: killing every processes with SIGTERM
2020-05-26T01:51:31.046+01:00
[WARNING] 146/005131 (1) : All workers exited. Exiting... (139)
2020-05-26T01:51:31.046+01:00
[ALERT] 145/233019 (1) : Current worker #1 (7) exited with code 143 (Terminated)
2020-05-26T00:30:19.662+01:00
[WARNING] 145/233019 (1) : All workers exited. Exiting... (0)
2020-05-26T00:30:19.662+01:00
[WARNING] 145/233019 (1) : Exiting Master process...
2020-05-26T00:30:19.636+01:00
```

I had two containers at the time to give redundancy in case of errors and funny enough, both containers crashed with the same error and around the same time: 2020-05-26T01:51:31.046+01:00 and 2020-05-26T01:53:44.661+01:00.

This is all I have on the logs.

And this is the problem with this bug:
<img width="828" alt="Screenshot 2020-05-26 at 07 51 29" src="https://user-images.githubusercontent.com/24032/82869178-b8210700-9f25-11ea-865a-e36e459671c1.png">

Since the application has only one worker there is no redundancy to failure in that case. If the worker dies for some reason the entire application starts getting 5XX errors.

Is there any way to work around this? Like having two workers or running it with multiple daemons?

**Comment by jiangwenyuan** [comment]
@igorescobar v5.0.10.21?

**Comment by igorescobar** [comment]
Yeah, that's the `stable` tag on Dockerhub.

**Comment by igorescobar** [comment]
I will retrieve the latest image again and re-deploy. I will keep you posted.

**Comment by jiangwenyuan** [comment]
Before you retrieve the lastest image, can you check the version of the crashed image?

**Comment by igorescobar** [comment]
Well, I guess now it's too late haha. But the command `docker pull nuster/nuster:stable` did retrieve a new image otherwise it would say that it is already up to date.

**Comment by igorescobar** [comment]
The current one is: `cd2271e36fd3`  167MB. and the previous one was: `69f38f828f2b` 95.8MB

**Comment by igorescobar** [comment]
Sorry, I got it wrong.
The current one is:
nuster/nuster                                                                 stable               9f556953a5ea        17 hours ago        95.8MB

**Comment by igorescobar** [comment]
The previous one might be:
nuster/nuster                                                                 <none>               69f38f828f2b        3 days ago          95.8MB

**Comment by jiangwenyuan** [comment]
And the one crashed was `69f38f828f2b` or `9f556953a5ea`?

**Comment by igorescobar** [comment]
Probably: `69f38f828f2b`. The `9f556953a5ea` is the one I just deployed.

**Comment by igorescobar** [comment]
It happened again with the latest stable image:
```
[NOTICE] 147/031811 (1) : New worker #1 (7) forked
2020-05-27T04:18:11.486+01:00
[ALERT] 147/031643 (1) : Current worker #1 (8) exited with code 139 (Segmentation fault)
2020-05-27T04:16:44.395+01:00
[ALERT] 147/031643 (1) : exit-on-failure: killing every processes with SIGTERM
2020-05-27T04:16:44.395+01:00
[WARNING] 147/031643 (1) : All workers exited. Exiting... (139)
2020-05-27T04:16:44.395+01:00
```

**Comment by jiangwenyuan** [comment]
@igorescobar Do you have a core file? Are you able to use gdb to catch the trace?

**Comment by igorescobar** [comment]
@jiangwenyuan I can't. Due to the auto-scaling, when this happens the container gets replaced and all the data gets lost.

**Comment by jiangwenyuan** [comment]
@igorescobar Tthe config file is similar with this one?

https://github.com/jiangwenyuan/nuster/issues/84#issuecomment-634535587

**Comment by igorescobar** [comment]
This is my config:
```
global
  maxconn 100000
  nuster cache on data-size 500m dict-size 500m dir /nuster/cache
  log stdout local0 info
  master-worker

  # lower your record size to improve Time to First Byte (TTFB)
  tune.ssl.maxrecord 1400

  # nuster manager on uri /internal/nuster
  # debug


defaults
  log global
  mode http

  log-format     '{"time":{"tr":%Tr,"tq":%Tq,"tw":%Tw,"tc":%Tc,"tt":%Tt},"haproxy":{"retries":%rc,"request":{"method":"%HM","uri":"%[capture.req.uri]","protocol":"%HV","header":{"host":"%[capture.req.hdr(0),json(utf8s)]","xforwardfor":"%[capture.req.hdr(1),json(utf8s)]","referer":"%[capture.req.hdr(2),json(utf8s)]"}},"name":{"server":"%s"},"response":{"status_code":%ST,"header":{"xrequestid":"%[capture.res.hdr(0),json(utf8s)]"}},"bytes":{"uploaded":%U,"read":%B}}}'
  option          dontlognull
  option          http-keep-alive
  option          http-server-close
  option          tcp-smart-accept
  option          tcp-smart-connect
  option          forwardfor header X-Real-IP

  retries 1
  option redispatch
  timeout http-keep-alive  140s
  timeout client           140s
  timeout connect          140s
  timeout server           140s

frontend fe
  bind *:3001
  option http-server-close
  default_backend be

backend be
  nuster cache on
  nuster rule r0 memory off disk on wait on key method.scheme.host.path ttl 12d if { res.hdr(Content-Type) -m beg image/ }
  http-response set-header x-cache HIT if { nuster.cache.hit }

  server s1 localhost:3000
```

**Comment by jiangwenyuan** [comment]
Ah, `log` is enabled. There is a `log` related segfault issue in haproxy. Not sure if it's related or not.

Let me try with this option. Meanwhile is it too bothered for you to disable log and see if it happens again?

**Comment by igorescobar** [comment]
@jiangwenyuan I can give it a try but the logs are pointing to stdout only, right?

**Comment by jiangwenyuan** [comment]
Yeah, thanks a lot

**Comment by igorescobar** [comment]
How would I know if it happened again the logs are off? 😅

**Comment by jiangwenyuan** [comment]
I see.. you are using container, and the container gets evicted if segfault happens..

Can we say if there's no segfault, there will be no eviction?

**Comment by igorescobar** [comment]
The container has health checks. When this happens the container stops responding on properly on the application port and ECS assumes the container needs to be replaced.

So as long as the /health_check route responds with a successful response the container won't be replaced.

**Comment by jiangwenyuan** [comment]
@igorescobar confirmed it's caused by `log` option.

**Comment by igorescobar** [comment]
Well, great you managed to reproduce it! At least now we know!

You mentioned an issue open on HAProxy project about this, can you share the link?

**Comment by jiangwenyuan** [comment]
https://github.com/haproxy/haproxy/issues/648

**Comment by igorescobar** [comment]
It just happened again. I will deploy it without the logs to see if I at least won't have the application unavailable.

**Comment by igorescobar** [comment]
Yeah, disabling logs fixes it.

**Comment by jiangwenyuan** [comment]
Should be fixed in haproxy v2.1.7 and in nuster v5.0.12.21
